### PR TITLE
Detailed fork tracker

### PR DESCRIPTION
**Pull request by gorbak25**
This PR maintains a list of key block hashes for which no successor is present - this also applies to orphan blocks.
The main use case is to immediately tell whether a large chain reorganization started and how much orphans are there.
Additionally this is a crucial step towards creating a GC for old orphaned blocks, switching between forks and mining on a particular fork one likes.
The loose ends are exposed via HTTP:
image
Frontends can then easily visualize orphans and mark orphaned transactions by using the new endpoint @thepiwo @mradkov 

Maintaining the loose end information is really cheap - everything is stored in the aec_chain_state table - I won't create another table due the fact that right now every table is backed by a fully independent database -.- When we finally make the full DB refactor then we can create a separate table and clean this up.

Old databases are also supported - If a old DB is detected then a DB sweep is started and we walk all key headers from genesis to the current top in order to find orphan blocks - this is done in the background and can be interrupted. For safety reasons if something in the migration crashes then we purposefully crash the entire node.

### CODE REVIEW

**Review by velzevur — APPROVED** [review]

**Comment by velzevur** [comment]
Since it is client visible, possibly worth the line in the release notes.

**Comment by gorbak25** [comment]
Before merging let me actually run the migration on a DB backup

File: apps/aecore/src/aec_db.erl [hunk 1]
```diff
@@ -556,6 +566,28 @@ write_top_block_hash(Hash) when is_binary(Hash) ->
...
+       [{aec_chain_state, {end_block_hash, Hash}}]).
+
+unmark_chain_end_hash(Hash) when is_binary(Hash) ->
+    ?t(mnesia:delete(aec_chain_state, {end_block_hash, Hash}, write)).
+
+find_chain_end_hashes() ->
+    [H || {end_block_hash, H} <- mnesia:dirty_all_keys(aec_chain_state)].
+
+write_chain_end_migration_state(State) ->
+    ?t(mnesia:write(#aec_chain_state{key = end_block_migration, value = State})). %% Crashes when error keys are present
```
[the remark this answers is not in the document]
Comment (uwiger) [reply apps/aecore/src/aec_db.erl]: So the successful write keeps the transaction from crashing?

[the remark this answers is not in the document]
Comment (gorbak25) [reply apps/aecore/src/aec_db.erl]: Personally I don't know what's going on - especially block insertion works fine so the error store should be already running.

**Comment by gorbak25** [comment]
Here are the results of my manual testing - I've ran the experiments on an c5d.2xlarge AWS instance :)
1. The migration seems to work fine - on the d4c372d45569e586b84d762ccb678fe7 snapshot it managed to complete in 20-30 minutes, but it did not find any orphan keyblocks which is concerning.
2. The migration is immune to killall -9 beam.smp and ./bin/aeternity stop - after restarting the migration is resumed as expected
3. Unfortunately the migration did not find a single fork on the d4c372d45569e586b84d762ccb678fe7 snapshot
4. I've let the node sync with the community fork and then with the attacker's fork - the chain_ends endpoint is displaying both forks as intended and even managed to find an orphaned gossiped block:
image

Now I'm waiting for the node to fully sync with the fork of the attacker - after which I will manually trigger the migration to see if the migration finds the forks.

**Comment by gorbak25** [comment]
Looks like the migration works. Now I will run the migration against 9e42014394090c0ea1f656975cbed7bf which should contain more forks.

**Comment by gorbak25** [comment]
The migration against 9e42014394090c0ea1f656975cbed7bf runs rather slow - the ETA of completion is 1-2h :( We could speed this up by bypassing mnesia and talking to rocksdb directly but I don't think this will be a good  idea.

**Comment by gorbak25** [comment]
Ok somehow 2 times in a row the migration crashed with:
```
2021-01-18 22:46:06.941 [warning] <0.5572.4>@aec_db:try_activity:1119 Mnesia activity Type=transaction exit with Reason={aborted,{{case_clause,{error,unknown_hash}},[{aec_chain_state,chain_ends_maybe_apply,3,[{file,"/home/builder/aeternity/apps/aecore/src/aec_chain_state.erl"},{line,1277}]},{aec_chain_state,'-chain_ends_finish_migration/2-fun-2-',1,[{file,"/home/builder/aeternity/apps/aecore/src/aec_chain_state.erl"},{line,1265}]},{aec_db,'-do_activity/2-fun-8-',1,[{file,"/home/builder/aeternity/apps/aecore/src/aec_db.erl"},{line,1137}]},{mnesia_tm,apply_fun,3,[{file,"mnesia_tm.erl"},{line,842}]},{mnesia_tm,execute_transaction,5,[{file,"mnesia_tm.erl"},{line,818}]},{mnesia,wrap_trans,6,[{file,"mnesia.erl"},{line,495}]},{aec_db,do_activity,2,[{file,"/home/builder/aeternity/apps/aecore/src/aec_db.erl"},{line,1136}]},{aec_db,try_activity,4,[{file,"/home/builder/aeternity/apps/aecore/src/aec_db.erl"},{line,1116}]}]}}, retrying
```
Invesigating

**Comment by gorbak25** [comment]
image

**Comment by gorbak25** [comment]
The test node node crashes on a block present in the fork of the attacker:
image

**Comment by gorbak25** [comment]
Ok, the migration against 9e42014394090c0ea1f656975cbed7bf works:
image
Let me optimize the migration slightly.

**Review by uwiger — DISMISSED** [review]

**Review by erlmachinedev — DISMISSED** [review]

**Comment by gorbak25** [comment]
The new migration works fine and executes in minutes even while syncing:
image

**Review by radrow — APPROVED** [review]

**Review by velzevur — APPROVED** [review]

**Review by dincho — COMMENTED** [review]
Could you please find a better place (url prefix) for this endpoint. Status is all about health checking, not data provider.

I would even consider having this under the internal endpoints if it's resource intensive.

**Review by uwiger — APPROVED** [review]

**Comment by gorbak25** [comment]
> Could you please find a better place (url prefix) for this endpoint. Status is all about health checking, not data provider.
> 
> I would even consider having this under the internal endpoints if it's resource intensive.
The endpoint is really cheap and there will be usually at most few entries. The chain ends are complementary to the status endpoint - it shows an overview of the chain. I agree that it's not the best place for it but it's also not that bad.

**Pull request merged**